CDS/src/org/icpc/tools/cds/CDSAuth.java

```diff
@@ -0,0 +1,146 @@
+
+import java.util.Base64;
+import java.util.List;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.security.enterprise.AuthenticationException;
+import javax.security.enterprise.AuthenticationStatus;
+import javax.security.enterprise.authentication.mechanism.http.HttpAuthenticationMechanism;
+import javax.security.enterprise.authentication.mechanism.http.HttpMessageContext;
+import javax.security.enterprise.credential.UsernamePasswordCredential;
+import javax.security.enterprise.identitystore.CredentialValidationResult;
+import javax.security.enterprise.identitystore.CredentialValidationResult.Status;
+import javax.security.enterprise.identitystore.IdentityStoreHandler;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+
+import org.icpc.tools.contest.Trace;
+import org.icpc.tools.contest.model.IAccount;
+
+@ApplicationScoped
+public class CDSAuth implements HttpAuthenticationMechanism {
+	private static final String AUTHORIZATION_HEADER = "Authorization";
+	private static final String BASIC_PREFIX = "Basic ";
+	private static final String RESULT = "result";
+	private static final String ACCOUNT = "account";
+
+	@Inject
+	private IdentityStoreHandler identityStoreHandler;
+
+	@Override
+	public AuthenticationStatus validateRequest(HttpServletRequest request, HttpServletResponse response,
+			HttpMessageContext context) throws AuthenticationException {
+		// check if the user is trying to login via form
+		String username = request.getParameter("username");
+		String password = request.getParameter("password");
+
+		if (username != null && password != null)
+			return validate(username, password, request, context);
+
+		// check if user is trying basic auth (but not for UI)
+		String uri = request.getRequestURI();
+		if (uri.startsWith("/api") || uri.startsWith("/presentation") || uri.startsWith("/stream")) {
+			String authHeader = request.getHeader(AUTHORIZATION_HEADER);
+			if (authHeader != null && authHeader.startsWith(BASIC_PREFIX)) {
+				String authValue = authHeader.substring(BASIC_PREFIX.length());
+				String userAndPassword = new String(Base64.getDecoder().decode(authValue));
+				if (!userAndPassword.contains(":"))
+					return context.responseUnauthorized();
+
+				username = userAndPassword.substring(0, userAndPassword.indexOf(':'));
+				password = userAndPassword.substring(userAndPassword.indexOf(':') + 1);
+
+				if (username != null && password != null)
+					return validate(username, password, request, context);
+			}
+		}
+
+		// check for host-based auto-login
+		if (request.getRemoteUser() == null) {
+			CDSConfig config = CDSConfig.getInstance();
+			username = config.getUserFromHost(request.getRemoteHost());
+			if (username == null)
+				username = config.getUserFromHost(request.getRemoteAddr());
+
+			if (username != null) {
+				password = config.getUserPassword(username);
```

Comment:
Could we (if we wanted) now drop the requirement that a password is required in the JSON if you do IP auth? I'm not too familiar with the Java auth stuff, but does the UsernamePasswordCredential below need the password? I guess it does?

Reply:
I *think* we could do that now since we control both identity store and auth, but decided to leave investigating that until later.